app/controllers/concerns/saml_idp_auth_concern.rb

```diff
@@ -49,17 +49,18 @@ def default_authn_context
 
   def link_identity_from_session_data
     IdentityLinker.new(
-      current_user,
-      current_issuer
-    ).link_identity(
-      ial: loa3_requested? ? 3 : 1
-    )
+      current_user, current_issuer
+    ).link_identity(ial: ial_level)
   end
 
   def identity_needs_verification?
     loa3_requested? && current_user.decorate.identity_not_verified?
   end
 
+  def ial_level
+    loa3_requested? && current_user.decorate.identity_verified? ? 3 : 1
+  end
```

Comment:
The fact that you had to add this to the SAML controller, but not the OIDC controller is an indication that something is not right.